pkgs/development/tools/misc/gdb/default.nix

```diff
@@ -156,6 +155,9 @@ stdenv.mkDerivation rec {
     license = lib.licenses.gpl3Plus;
 
     platforms = with lib.platforms; linux ++ cygwin ++ freebsd ++ darwin;
+    # upstream does not support targeting aarch64-darwin;
+    # see https://inbox.sourceware.org/gdb/3185c3b8-8a91-4beb-a5d5-9db6afb93713@Spark/
+    badPlatforms = lib.optionals (stdenv.targetPlatform.system == "aarch64-darwin") meta.platforms;
```

Comment:
Maybe `[ stdenv.hostPlatform.system ]` would make more sense than `meta.platforms` here? Not sure, it’s kind of a weird construction…

Reply:
It seems both ([1](https://github.com/NixOS/nixpkgs/blob/b79ce4c43f9117b2912e7dbc68ccae4539259dda/pkgs/development/cuda-modules/tensorrt/fixup.nix#L110), [2](https://github.com/NixOS/nixpkgs/blob/b79ce4c43f9117b2912e7dbc68ccae4539259dda/pkgs/development/cuda-modules/setup-hooks/extension.nix#L51)) are used; either way this isn't communicating to the user that the _host_ platform has nothing to do with it, and it's a problem with the _target_ platform. I'm partial to your suggestion because it's clearer when trying to build `gdb` without cross compilation on `aarch64-darwin`. However, that implementation is not correct on certain edge cases (let me stress I can't imagine actually doing this, but still):
```
$ nix-instantiate --eval --expr 'with import ./. { system="aarch64-darwin"; crossSystem="aarch64-darwin"; }; lib.meta.availableOn { system = "x86_64-linux"; } gdb'
true # incorrect
```

In any case, I'm not satisfied with the lack of explanation given to the user, but I couldn't find a way to defer evaluation of a `warn` to only execute if the `meta` check actually failed. Any ideas?

Reply:
Yeah, I’m not sure about a warning. `badPlatforms = lib.optionals (stdenv.targetPlatform.system == "aarch64-darwin") (lib.warn …)` would I guess produce failures just from an `availableOn` check. It might be impossible to get something truly nice here since the problem is that the `platforms` machinery fails to encode all the possible dimensions. `broken` is less strictly accurate than `badPlatforms`, but would avoid having to pick a value to conditionally mark as “bad”, and the error message may be less misleading merely by virtue of being more vague. `availableOn` doesn’t check `broken`, though, so… I’m not sure I have a better idea than the two we’ve already come up with. I’m not super knowledgable about cross in general, so it’s possible I’m missing something obvious. (I can’t really figure out why the example you gave fails with my suggestion, for instance!)

Reply:
Whenever the `gdb` package gets evaluated, platforms in `meta` will be checked against `stdenv.hostPlatform`, and `stdenv.hostPlatform == stdenv.hostPlatform`: life is good.
https://github.com/NixOS/nixpkgs/blob/40222c1c96c73fdb47070417a5ae34ffb314ad35/pkgs/stdenv/generic/check-meta.nix#L104-L105

But `lib.availableOn` can take any `system` as an argument, which does not necessarily have to be `stdenv.hostPlatform` (although I can't imagine *why* you would want them to be different). Semantically, if `targetPlatform.system == "aarch64-darwin"`, _there is no good `hostPlatform`_ to build `gdb` on: they would all fail.

Maybe `meta.broken` would be less confusing to end users, even though nothing is technically broken (unless a lack of support is considered broken). Either way, I think they would have to consult nixpkgs source to see why evaluation fails. I don't have enough nixpkgs experience to determine what would be best for package maintainers and users, so I'll defer to your judgement.

Reply:
I think this is fine as it is, in lieu of a `meta` that actually understands cross. If someone else can think of something better than that could be handled in a follow‐up. Thanks for handling the fix!